elasticsearch/_async/http_aiohttp.py

```diff
@@ -0,0 +1,189 @@
+            cloud_id=cloud_id,
+            api_key=api_key,
+            opaque_id=opaque_id,
+            **kwargs,
+        )
+
+        if http_auth is not None:
+            if isinstance(http_auth, str):
+                http_auth = tuple(http_auth.split(":", 1))
+
+            if isinstance(http_auth, (tuple, list)):
+                http_auth = aiohttp.BasicAuth(*http_auth)
+
+        # if providing an SSL context, raise error if any other SSL related flag is used
+        if ssl_context and (
+            (verify_certs is not VERIFY_CERTS_DEFAULT)
+            or ca_certs
+            or client_cert
+            or client_key
+            or ssl_version
+        ):
+            warnings.warn(
+                "When using `ssl_context`, all other SSL related kwargs are ignored"
+            )
+
+        self.ssl_assert_fingerprint = ssl_assert_fingerprint
+        if self.use_ssl and ssl_context is None:
+            ssl_context = ssl.SSLContext(ssl_version or ssl.PROTOCOL_TLS)
+
+            # Convert all sentinel values to their actual default
+            # values if not using an SSLContext.
+            if verify_certs is VERIFY_CERTS_DEFAULT:
+                verify_certs = True
+
+            ca_certs = CA_CERTS if ca_certs is None else ca_certs
+            if verify_certs:
+                if not ca_certs:
+                    raise ImproperlyConfigured(
+                        "Root certificates are missing for certificate "
+                        "validation. Either pass them in using the ca_certs parameter or "
+                        "install certifi to use it automatically."
+                    )
+            if os.path.isfile(ca_certs):
+                ssl_context.load_verify_locations(cafile=ca_certs)
+            elif os.path.isdir(ca_certs):
+                ssl_context.load_verify_locations(capath=ca_certs)
+            else:
+                raise ImproperlyConfigured("ca_certs parameter is not a path")
+
+        self.headers.setdefault("connection", "keep-alive")
+        self.session = aiohttp.ClientSession(
+            auth=http_auth,
+            headers=self.headers,
+            auto_decompress=True,
+            connector=aiohttp.TCPConnector(
+                limit=maxsize,
+                verify_ssl=verify_certs,
+                use_dns_cache=True,
+                ssl_context=ssl_context,
+                keepalive_timeout=10,
+            ),
+        )
+
+    async def close(self):
+        await self.session.close()
+
+    async def perform_request(
+        self, method, url, params=None, body=None, timeout=None, ignore=(), headers=None
+    ):
+        url_path = url
+        if params:
+            url_path = "%s?%s" % (url, urlencode(params or {}))
+        url = self.host + url_path
+
+        timeout = aiohttp.ClientTimeout(total=timeout)
+        req_headers = self.headers.copy()
+        if headers:
+            req_headers.update(headers)
+
+        loop = get_running_loop()
```

Comment:
I think this can cause a performance penalty and I wonder whether it is worth caching the eventloop in a field (based on your PR description I understand you want to retrieve it lazily). The problem is that [the CPython implementation](https://github.com/python/cpython/blob/df8913f7c48d267efd662e8ffd9496595115eee8/Lib/asyncio/events.py#L693-L702) calls `getpid` on every call which in turn will [issue the `getpid` system call](http://man7.org/linux/man-pages/man2/getpid.2.html) for glibc > 2.25 (also [Python does not cache the pid](https://github.com/python/cpython/blob/2b5603140c09766a7d4e8243a70d7144f684f6f9/Modules/posixmodule.c#L6926-L6939)).

Reply:
I'll cache this along with the "sniff on first async action", good point!